Applications/Tizen_native/CustomShortcut/src/data.c

```diff
@@ -43,6 +43,199 @@ int parse_route(const char *source, char **route, char **data) {
+
+  void *raw_data;
+  size_t data_size;
+  FILE *file;
+
+  pthread_mutex_lock(&ad->pipe_lock);
+
+  /// @note data_size written here will be overriden and it is intended.
+  int status = ml_tensors_data_get_tensor_data(data, 0, &raw_data, &data_size);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("get tensor data failed %d", status);
+    status = ml_tensors_info_get_tensor_size(info, -1, &data_size);
+    goto CLEAN;
+  }
+
+  file = fopen(ad->pipe_dst, ad->tries == 1 ? "wb" : "wb+");
+
+  if (file == NULL) {
+    LOG_E("cannot open file");
+  }
+
+  if (fwrite(raw_data, data_size, 1, file) < 0) {
+    LOG_E("write error happend");
+  };
+
+  if (fclose(file) < 0) {
+    LOG_E("there was error closing");
+    goto CLEAN;
+  }
+
+  LOG_D("using pipeline finished, destroying pipeline");
+
+CLEAN:
+  pthread_cond_signal(&ad->pipe_cond);
+  pthread_mutex_unlock(&ad->pipe_lock);
+}
+
+static int _run_nnpipeline(appdata_s *ad, const char *src, bool append) {
+  char pipe_description[5000];
+
+  char model_path[PATH_MAX];
+
+  int status = ML_ERROR_NONE;
+
+  data_get_resource_path("mobilenetv2.tflite", model_path, false);
+
+  status = pthread_mutex_lock(&ad->pipe_lock);
+  if (status != 0) {
+    LOG_E("acquiring lock failed status: %d", status);
+    return status;
+  }
+
+  LOG_D("pipe ready, starting pipeline");
+
+  sprintf(pipe_description,
+          "filesrc location=%s ! pngdec ! "
+          "videoconvert ! videoscale ! "
+          "video/x-raw,width=224,height=224,format=RGB ! "
+          "tensor_converter ! "
+          "tensor_transform mode=arithmetic option=%s ! "
+          "tensor_filter framework=tensorflow-lite model=%s ! "
+          "tensor_sink name=sink",
+          src, "typecast:float32,add:-127.5,div:127.5", model_path);
+
+  LOG_D("setting inference \n pipe: %s", pipe_description);
+  status = ml_pipeline_construct(pipe_description, NULL, NULL, &ad->pipeline);
+
+  if (status != ML_ERROR_NONE) {
+    LOG_E("something wrong constructing pipeline %d", status);
+    ml_pipeline_destroy(ad->pipeline);
+    pthread_mutex_unlock(&ad->pipe_lock);
+    return status;
+  }
+
+  status = ml_pipeline_sink_register(ad->pipeline, "sink", _on_data_receive,
+                                     (void *)ad, &ad->pipe_sink);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("sink register failed %d", status);
+    ml_pipeline_destroy(ad->pipeline);
+    pthread_mutex_unlock(&ad->pipe_lock);
+    return status;
+  }
+
+  LOG_D("starting inference");
+  status = ml_pipeline_start(ad->pipeline);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("failed to start pipeline %d", status);
+    ml_pipeline_destroy(ad->pipeline);
+    pthread_mutex_unlock(&ad->pipe_lock);
+    return status;
+  }
+
+  pthread_cond_wait(&ad->pipe_cond, &ad->pipe_lock);
+
+  LOG_D("stopping pipeline");
+  status = ml_pipeline_stop(ad->pipeline);
+  if (status != ML_ERROR_NONE) {
+    LOG_E("stopping pipeline failed");
+    ml_pipeline_destroy(ad->pipeline);
```

Comment:
ditto.

Reply:
you can use goto for easier cleanup here.

Reply:
changed to cleanup